plugins/logitech-bulkcontroller/fu-plugin-logitech-bulkcontroller.c

```diff
@@ -0,0 +1,388 @@
+			  plugin_data->device_name,
+			  bulk_intf);
+		g_cond_signal(&plugin_data->test_upd_cond);
+		break;
+	case TRANSFER_INPROGRESS:
+		if (BULK_INTERFACE_UPD == bulk_intf) {
+			FuDevice *device;
+			g_debug("\r[%s] : Interface: %u File uploaded : [%d%%]",
+				plugin_data->device_name,
+				bulk_intf,
+				progress);
+			fflush(stdout);
+			device = plugin_data->device_obj;
+			if (device)
+				fu_device_set_progress(plugin_data->device_obj, progress);
+		} else if (BULK_INTERFACE_SYNC == bulk_intf) {
+			g_debug("\r[%s] : Interface: %u File received : [%d%%]",
+				plugin_data->device_name,
+				bulk_intf,
+				progress);
+			fflush(stdout);
+		}
+		break;
+	case TRANSFER_COMPLETED:
+		g_debug("[%s] :  File transfer completed for interface %u",
+			plugin_data->device_name,
+			bulk_intf);
+		g_cond_signal(&plugin_data->test_upd_cond);
+		break;
+	default:
+		break;
+	}
+}
+
+gboolean
+send_device_info_request(void *data)
+{
+	gint ret;
+	Message message = {0};
+	FuPluginData *plugin_data;
+	plugin_data = (FuPluginData *)data;
+	g_cond_init(&plugin_data->test_upd_cond);
+	g_mutex_init(&plugin_data->test_upd_mutex);
+	g_mutex_lock(&plugin_data->test_upd_mutex);
+
+	ret = proto_manager_generate_get_device_info_request(&message);
+	if (!ret && message.data) {
+		ReturnValue *ret_val;
+		g_debug("Sending GetDeviceInfoRequest. Length %lu", message.len);
+		ret_val = logibulkcontroller_send_data_sync(plugin_data->bulkcontroller_obj,
+							    message.data,
+							    message.len);
+		g_free(message.data);
+		if ((ERRORCODE_NO_ERROR != ret_val->error_code) &&
+		    (ERRORCODE_SEND_DATA_REQUEST_PUSHED_TO_QUEUE != ret_val->error_code)) {
+			g_warning("Error in sending GetDeviceInfoRequest %u", ret_val->error_code);
+			return FALSE;
+		}
+		if (g_cond_wait_until(&plugin_data->test_upd_cond,
+				      &plugin_data->test_upd_mutex,
+				      g_get_monotonic_time() + 5 * G_TIME_SPAN_SECOND)) {
+			g_mutex_unlock(&plugin_data->test_upd_mutex);
+			g_mutex_lock(&plugin_data->test_upd_mutex);
+		}
+	}
+	return TRUE;
+}
+
+void
+fu_plugin_init(FuPlugin *plugin)
+{
+	g_debug("Initializing Logitech bulk controller plugin");
+	fu_plugin_set_build_hash(plugin, FU_BUILD_HASH);
+	fu_plugin_alloc_data(plugin, sizeof(FuPluginData));
+}
+
+gboolean
+fu_plugin_coldplug(FuPlugin *plugin, GError **error)
+{
+	LogiBulkController *obj;
+	FuPluginData *plugin_data;
+	gint pid;
+	FuContext *ctx;
+	BulkControllerCallbacks bulkcb = {bulk_error_cb,
+					  bulk_file_transfer_cb,
+					  read_upd_data_cb,
+					  read_sync_data_cb,
+					  send_data_sync_cb};
+
+	plugin_data = fu_plugin_get_data(plugin);
+	plugin_data->bulkcontroller_obj = NULL;
+	/* Check if Logitech Rally Bar Mini is conneced */
+	pid = 0x08D3;
+	plugin_data->device_name = g_strdup("Rally Bar Mini");
+	plugin_data->device_type = kDeviceTypeRallyBarMini;
+	plugin_data->device_info_available = FALSE;
+	plugin_data->run_devicemode_info_ = FALSE;
```

Comment:
as a general statement - it feels very much so that this plugin is architected to only handle one device at a time, but given these are USB there is no reason that someone can't connect two simultaneously.  The way it's built right now it seems like only one would ever be accessible.

Rather than creating private plugin data, I think you should be subclassing a device and having device private data.

Reply:
I'll have a go at splitting this up this morning.